### PR TITLE
Refactor(keymaps): export keymaps

### LINKED ISSUE #149: [FEATURE] Allow manual completion via function
**Is your feature request related to a problem? Please describe.**
When using this plugin, it is recommended that we map the `accept_key` to something that doesn't collide with other plugin mappings. While this works, it would be interesting if we can do a manual complete ourselves by using some kind of provided function. Generally, autocomplete engines provide ways of doing this ([asyncomple.vim](https://github.com/prabirshrestha/asyncomplete.vim/blob/master/doc/asyncomplete.txt#L129), [ddc.vim](https://github.com/Shougo/ddc.vim/blob/main/doc/ddc.txt#L531), [ncm2](https://github.com/ncm2/ncm2/blob/master/doc/ncm2.txt#L413), [CoC](https://github.com/neoclide/coc.nvim/blob/master/doc/coc.txt#L2203), [nvim-cmp](https://github.com/hrsh7th/nvim-cmp/blob/main/doc/cmp.txt#L247) etc).

**Describe the solution you'd like**
I'm going to type vimscript since I don't know how to do it in lua in neovim (migrating atm):
```vim
inoremap <silent> <expr> <cr>
    \ tabnine#has_suggestion() ?
    \ tabnine#confirm() : '<cr>'

```
**Describe alternatives you've considered**
The only alternative I've considered is having some function as `accept_key` that we can have logic attached to it that decides whether or not we want to return a valid keybind or just `nil`.

**Additional context**
Although this plugin doesn't serve as a autocomplete plugin, it very much has autocomplete functionality. Having an API that allows us to inspect the results of this plugin could even benefit other users of it; [cmp-tabnine](https://github.com/tzachar/cmp-tabnine) comes to mind.

### PR DESCRIPTION
**Pull request by aarondill**
This PR implements `accept_suggestion`, `dismiss_suggestion`, and `has_suggestion` to be used from user mappings.

It also permits setting `accept_keymap` and `dismiss_keymap` to `false` to avoid setting it (`nil` defaults to the default settings)

fixes #149 

I've included an example use case under a new header in the README.

copied from the readme:
```lua
--- Example integration with Tabnine and LuaSnip; falling back to inserting tab if neither has a completion
vim.keymap.set("i", "<tab>", function()
  if require("tabnine.keymaps").has_suggestion() then
    return require("tabnine.keymaps").accept_suggestion()
  elseif require("luasnip").jumpable(1) then
    return require("luasnip").jump(1)
  else
    return "<tab>"
  end
end, { expr = true })
```

### CODE REVIEW

**Comment by amirbilu** [comment]
Hi @aarondill 
Sorry about the delay. Super busy week. Will try to get it by next week.

Did you get the swag already 😎?

**Comment by aarondill** [comment]
> Hi @aarondill Sorry about the delay. Super busy week. Will try to get it by next week.
> 
> Did you get the swag already sunglasses?

No problem, take your time.

I did! I sent a few images to your email address already, did you not receive them?

**Comment by amirbilu** [comment]
I haven't! to `amirbilu@tabnine` or to `support@tabnine`? If you have a `twitter` handle - let us know so we can share your images & story at https://twitter.com/tabnine

**Comment by aarondill** [comment]
@amirbilu any word on this PR? I'm still thinking perhaps exporting these under `tabnine` rather than `tabnine.keymaps`, but i'd like your input on that, as well as their final names.

**Comment by aarondill** [comment]
@amirbilu what are your thoughts on this patch? I can say I've been happily using this in my own config for a month with no issues (see [here](https://github.com/aarondill/LazyVim/blob/8847f49f2d4958167d4063913c03da675dfc1a0c/lua/plugins/tabnine.lua#L75-L83)), and would love to get it merged into the main repo. I'd still like your input on the module name, whether `tabnine.keymaps.accept_completion()` or else `tabnine.accept_completion()`, or perhaps even just `tabnine.accept()`, and same for dismiss.

**Comment by aarondill** [comment]
@amirbilu please let me know how you would like this change to continue. This is vital to my personal use of the project, and I'd like to finalize the PR so it can be merged.

**Comment by aarondill** [comment]
@amirbilu have you had a chance to take a look at this yet?

**Comment by aarondill** [comment]
@amirbilu sorry if it feels like I'm pestering you about this PR, but this is an extremely important use case to me, and as you seem to be the sole owner/maintainer of this repo, I don't know who else to ping to get this reviewed and merged.

I am currently maintaining a fork to implement this behavior for my own use and I'd love to be able to resume using upstream if this could get merged.

My stated use case is to remap the tab key (usually used for completion) to a function callback that completes Tabnine if available, otherwise uses cmp (my lsp completion client). This allows seemless integration of the two, and avoids my current problem of the two fighting over my tab key (and Tabnine winning, regardless of completion availability, removing my ability to use cmp).

**Comment by aarondill** [comment]
@amirbilu here's a friendly ping on this PR. It's now been waiting over 6 months for a review and I'd appreciate if you could take a quick look at it.

**Comment by aarondill** [comment]
@amirbilu any update on this PR?

**Comment by amirbilu** [comment]
@aarondill to be honest it feels like a big change and I can't find the time to think of the consequences

**Comment by aarondill** [comment]
@amirbilu this pr would also solve #149 with slight modification (which i'm happy to make)

if I make these modifications, would you be interested in merging this PR now?

File: lua/tabnine/completion.lua [hunk 1]
```diff
@@ -13,6 +13,8 @@ local function valid_response(response)
 end
 
 function M.accept()
+	if not state.completions_cache or not state.rendered_completion then return end
```
Comment (amirbilu) [inline lua/tabnine/completion.lua:16]: can you explain why this is needed

**Comment by amirbilu** [comment]
> @amirbilu this pr would also solve #149 with slight modification (which i'm happy to make)
> 
> if I make these modifications, would you be interested in merging this PR now?

willing to add this if it does not break anything

File: lua/tabnine/completion.lua (hunk 1 above)
Comment (aarondill) [reply lua/tabnine/completion.lua]: This was needed to make sure the `utils.starts_with` didn't throw with an empty string.

The check has been rewritten to be more clear in the latest commits. Please see the new (rewritten) changes.

**Comment by aarondill** [comment]
@amirbilu I've rewritten this PR with `has_suggestion` in mind.
Please review the new commits when possible :smile:

**Review by amirbilu — APPROVED** [review]

**Comment by amirbilu** [comment]
Awesome contribution

**Pull request merged**